[WIP] Add support for dualstack / multiple IP ranges

Ayush21298 opened the pull request:
# New Feature

## Motivation

This PR adds support for IPv4/IPv6 dual-stack and multiple IP assignment

Feature proposal: #237 

## Changes

- Introduced a new type `RangeConfiguration` in `pkg/types/types.go` for encapsulating configurations of range for IP assignment
- Add new field `IPRanges  []RangeConfiguration` in `IPAMConfig` for handling multiple IP ranges
- Update `config.go` to work with `IPRanges` instead of `IPAM.Range, ...`
- Use `RangeConfiguration` in `allocate.AssignIP` instead of `IPAMConfig`
- Integrate `IPRanges` with etcd's IP Management
- Integrate `IPRanges` with kubernetes' IP Management
- Update whereabouts' logic to assign multiple new IPs to a pod based on a list of new IPs returned by IP Management module
- Update function definition of `garbageCollector` to match with the changes in definition of `wbclient.IPManagement`

## Tasks

### Project management

- [x] Created a related issue
- [ ] Added proper labels to the issue @dougbtv 
- [ ] Assigned yourself as the assignee @dougbtv 
- [x] Created a PR for design proposal
- [ ] Finalized and merged design proposal @dougbtv @maiqueb 

### Code management

- [x] Update IPAM structure
- [ ] Backward compatibility
- [x] Add feature to support dual-stack
- [x] Add feature to support multiple IP ranges
- [x] Manual testing for above feature
- [ ] Update old test cases
- [ ] Add new test cases for the added features
- [ ] Update documentation
- [x] Checked for mistakenly committed files

Closes #212

Ayush21298 commented:
# Test for multiple IP assignment

## NAD

```
apiVersion: "k8s.cni.cncf.io/v1"
kind: NetworkAttachmentDefinition
metadata:
  name: whereabouts-conf
spec:
  config: '{
      "cniVersion": "0.3.0",
      "name": "whereaboutsexample",
      "type": "macvlan",
      "master": "eth0",
      "mode": "bridge",
      "ipam": {
        "type": "whereabouts",
        "ipRanges": [{
          "range": "10.10.10.10/24",
          "range_start": "10.10.10.20",
          "range_end": "10.10.10.200"
        }, {
          "range": "11.11.11.11/24",
          "range_start": "11.11.11.20",
          "range_end": "11.11.11.200"
        }]
      }
    }'
```

## Pod IP Config

```
5: net1@if2: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default 
    link/ether aa:ac:37:3a:1a:91 brd ff:ff:ff:ff:ff:ff link-netnsid 0
    inet 10.10.10.20/24 brd 10.10.10.255 scope global net1
       valid_lft forever preferred_lft forever
    inet 11.11.11.20/24 brd 11.11.11.255 scope global net1
       valid_lft forever preferred_lft forever
    inet6 fe80::a8ac:37ff:fe3a:1a91/64 scope link 
       valid_lft forever preferred_lft forever
```

Ayush21298 commented:
# Test for IPv4/IPv6 dual-stack
## NAD

```
apiVersion: "k8s.cni.cncf.io/v1"
kind: NetworkAttachmentDefinition
metadata:
  name: whereabouts-conf
spec:
  config: '{
      "cniVersion": "0.3.0",
      "name": "whereaboutsexample",
      "type": "macvlan",
      "master": "eth0",
      "mode": "bridge",
      "ipam": {
        "type": "whereabouts",
        "ipRanges": [{
          "range": "10.10.10.10/24",
          "range_start": "10.10.10.20",
          "range_end": "10.10.10.200"
        }, {
          "range": "2001::1/64",
          "range_start": "2001::16",
          "range_end": "2001::64"
        }]
      }
    }'
```

## Pod IP Config

```
5: net1@if2: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default 
    link/ether ee:f1:0a:a4:f3:cd brd ff:ff:ff:ff:ff:ff link-netnsid 0
    inet 10.10.10.20/24 brd 10.10.10.255 scope global net1
       valid_lft forever preferred_lft forever
    inet6 2001::16/64 scope global 
       valid_lft forever preferred_lft forever
    inet6 fe80::ecf1:aff:fea4:f3cd/64 scope link 
       valid_lft forever preferred_lft forever
```

hhhybb commented:
@Ayush21298 Hi, is there any plan to support dual stack in the recent release?

Ayush21298 commented:
> @Ayush21298 Hi, is there any plan to support dual stack in the recent release?

@hhhybb Hi, I'll require some time to proceed with fixing some minor issues, adding backward compatibility and testing. If backward compatibility is not a major requirement, you can use this branch for basic DualStack functionality :)

hhhybb commented:
> > @Ayush21298 Hi, is there any plan to support dual stack in the recent release?
> 
> @hhhybb Hi, I'll require some time to proceed with fixing some minor issues, adding backward compatibility and testing. If backward compatibility is not a major requirement, you can use this branch for basic DualStack functionality :)

@Ayush21298 Thanks for your reply, I have installed the old versions in the production environment and need backward compatibility when upgrading. I am looking forward to it can be merged into master. Thanks for your efforts.

Ayush21298 commented:
# Test for IPv4/IPv6 dual-stack along with Old Range configuration

## NAD

```
apiVersion: "k8s.cni.cncf.io/v1"
kind: NetworkAttachmentDefinition
metadata:
  name: whereabouts-conf
spec:
  config: '{
      "cniVersion": "0.3.0",
      "name": "whereaboutsexample",
      "type": "macvlan",
      "master": "eth0",
      "mode": "bridge",
      "ipam": {
        "type": "whereabouts",
        "range": "12.12.12.12/24",
        "ipRanges": [{
          "range": "10.10.10.10/24",
          "range_start": "10.10.10.20",
          "range_end": "10.10.10.200"
        }, {
          "range": "2001::1/64",
          "range_start": "2001::16",
          "range_end": "2001::64"
        }]
      }
    }'
```

## Pod IP Config
```
5: net1@if2: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default 
    link/ether 8e:58:0a:33:88:10 brd ff:ff:ff:ff:ff:ff link-netnsid 0
    inet 10.10.10.20/24 brd 10.10.10.255 scope global net1
       valid_lft forever preferred_lft forever
    inet 12.12.12.1/24 brd 12.12.12.255 scope global net1
       valid_lft forever preferred_lft forever
    inet6 2001::16/64 scope global 
       valid_lft forever preferred_lft forever
    inet6 fe80::8c58:aff:fe33:8810/64 scope link 
       valid_lft forever preferred_lft forever
```

maiqueb commented:
> # Test for IPv4/IPv6 dual-stack along with Old Range configuration
> ## NAD
> ```
> apiVersion: "k8s.cni.cncf.io/v1"
> kind: NetworkAttachmentDefinition
> metadata:
>   name: whereabouts-conf
> spec:
>   config: '{
>       "cniVersion": "0.3.0",
>       "name": "whereaboutsexample",
>       "type": "macvlan",
>       "master": "eth0",
>       "mode": "bridge",
>       "ipam": {
>         "type": "whereabouts",
>         "range": "12.12.12.12/24",
>         "ipRanges": [{
>           "range": "10.10.10.10/24",
>           "range_start": "10.10.10.20",
>           "range_end": "10.10.10.200"
>         }, {
>           "range": "2001::1/64",
>           "range_start": "2001::16",
>           "range_end": "2001::64"
>         }]
>       }
>     }'
> ```
> 
> ## Pod IP Config
> ```
> 5: net1@if2: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default 
>     link/ether 8e:58:0a:33:88:10 brd ff:ff:ff:ff:ff:ff link-netnsid 0
>     inet 10.10.10.20/24 brd 10.10.10.255 scope global net1
>        valid_lft forever preferred_lft forever
>     inet 12.12.12.1/24 brd 12.12.12.255 scope global net1
>        valid_lft forever preferred_lft forever
>     inet6 2001::16/64 scope global 
>        valid_lft forever preferred_lft forever
>     inet6 fe80::8c58:aff:fe33:8810/64 scope link 
>        valid_lft forever preferred_lft forever
> ```

Cool. Please add an e2e test with this behavior, by introducing a new context in [here](https://github.com/k8snetworkplumbingwg/whereabouts/blob/8874106d77e854a2fc566737350bfe3eee3d099d/e2e/e2e_test.go#L87).

Ayush21298 commented:
> Cool. Please add an e2e test with this behavior, by introducing a new context in [here](https://github.com/k8snetworkplumbingwg/whereabouts/blob/8874106d77e854a2fc566737350bfe3eee3d099d/e2e/e2e_test.go#L87).

Sure :)
I wanted to do that before the meeting, but got a bit delayed :sweat_smile:

dougbtv commented on pkg/config/config.go:
```diff
@@ -76,30 +76,47 @@ func LoadIPAMConfig(bytes []byte, envArgs string, extraConfigPaths ...string) (*
-			n.IPAM.RangeStart = firstip
+  if n.IPAM.Range != "" {
+
+    logging.Debugf("Deprecated Range present in IPAM config. It will soon be removed in later releases. Please use IPRanges instead")
```
I'm not sure about actually deprecating this. I think I like having a single range still be an option, and it's simpler to type in. However, at debugf logging level, I think I'm OK with this comment. (no change necessary, thinking out loud)

Ayush21298 replied:
> I'm not sure about actually deprecating this. I think I like having a single range still be an option, and it's simpler to type in. However, at debugf logging level, I think I'm OK with this comment. (no change necessary, thinking out loud)

@dougbtv Thank you so much for reviewing :smiley:

Ayush21298 commented:
> Cool. Please add an e2e test with this behavior, by introducing a new context in [here](https://github.com/k8snetworkplumbingwg/whereabouts/blob/8874106d77e854a2fc566737350bfe3eee3d099d/e2e/e2e_test.go#L87).

@maiqueb I have added both e2e test cases as well as unit tests for whereabouts' module :)
Please review them when you are available ^^

kmlebedev commented:
@dougbtv Hi, I would like to start using multiple IP ranges right now, judging by the comments there is already a working version of this feature https://github.com/k8snetworkplumbingwg/whereabouts/pull/60
Can someone confirm that this PR is already in use by someone and is it stable?

dougbtv commented:
@kmlebedev -- #60 was rejected (my apologies to the contributor) so there's no guarantee of compatibility, my apologies but it was never in master. So you might need to figure out your own migration path.

Ayush21298 commented:
> #225 is merged @Ayush21298 we're good for a rebase, thanks!

@dougbtv Thanks for letting me know.
I'll notify you after rebasing :)

Ayush21298 commented:
@dougbtv @maiqueb I have rebased the code :)
Can you please have a look ?

maiqueb commented on pkg/config/config.go:
```diff
@@ -112,7 +112,7 @@ func LoadIPAMConfig(bytes []byte, envArgs string, extraConfigPaths ...string) (*
 				return nil, "", fmt.Errorf("invalid CIDR %s: %s", n.IPAM.IPRanges[idx].Range, err)
 			}
-			n.IPAM.Range = ipNet.String()
+			n.IPAM.IPRanges[idx].Range = ipNet.String()
```
The unit tests should be fixed in the same commit where the production code has changed. 

Doing it like as you're proposing would render tools like `git bisect` useless, and also make things a lot harder for the reviewers.

This commit should be squashed with the commit that added the range configuration to the `IPAMConfig` struct.

Ayush21298 replied:
Done :+1:

maiqueb commented on e2e/e2e_test.go:
```diff
@@ -73,7 +73,7 @@ var _ = Describe("Whereabouts functionality", func() {
 			Expect(err).NotTo(HaveOccurred())
 
-			netAttachDef = macvlanNetworkWithWhereaboutsIPAMNetwork(testNetworkName, testNamespace, ipv4TestRange)
+			netAttachDef = macvlanNetworkWithWhereaboutsIPAMNetwork(testNetworkName, testNamespace, ipv4TestRange, []string{})
```
I don't understand this commit - it does nothing. 

It changes the test code to accept another parameter, but parameter is always passed empty. 

Please squash the commit that makes use of this so I can understand how this is used.

Ayush21298 replied:
Done :+1:

maiqueb commented on e2e/e2e_test.go:
```diff
@@ -106,9 +112,84 @@ var _ = Describe("Whereabouts functionality", func() {
+			})
+
+			AfterEach(func() {
+				By("deleting pod with whereabouts net-attach-def")
+				Expect(clientInfo.DeletePod(pod)).To(Succeed())
+				By("deleting DualStack NetworkAttachmentDefinition for whereabouts")
+				Expect(clientInfo.DelNetAttachDef(netAttachDefDualStack)).To(Succeed())
```
I would remove the `By` clauses. 

They're pretty much useless here - you just want to get rid of existing entities ... What is the motivation for having these ?

Ayush21298 replied:
I was trying to keep the coding style consistent with other tests. For e.g.

https://github.com/k8snetworkplumbingwg/whereabouts/blob/38ded4bbc37e5205f4cefc75c6902fa414ef32d2/e2e/e2e_test.go#L109-L110

maiqueb commented on cmd/whereabouts_test.go:
```diff
@@ -448,6 +448,164 @@ var _ = Describe("Whereabouts operations", func() {
 
 	})
 
+	It("allocates an address using IPRanges notation", func() {
```
please squash these unit tests to the commit that changed the production code.

Ayush21298 replied:
Done :+1:

maiqueb commented on cmd/whereabouts_test.go:
```diff
@@ -557,12 +557,16 @@ var _ = Describe("Whereabouts operations", func() {
 			IfName:      ifname,
 			StdinData:   []byte(conf),
-			Args:        "IgnoreUnknown=1;K8S_POD_NAMESPACE=dummyNS;K8S_POD_NAME=dummyPOD",
+			Args:        cniArgs(podNamespace, podName),
```
I don't understand this commit at all. This code already exists in the master branch. You should rebase your code instead of updating this manually.

Ayush21298 replied:
It is part of the new test introduced in this PR so it didn't exist in the master branch.

https://github.com/k8snetworkplumbingwg/whereabouts/blob/310ad3d8ae10e8d199d6c62aaa23c21d16f223b0/cmd/whereabouts_test.go#L534

This was added later on while updating the tests based on new changes merged in the master branch.
It will be squashed with the commit of this test :)

maiqueb commented on pkg/config/config.go:
```diff
@@ -120,6 +120,15 @@ func LoadIPAMConfig(bytes []byte, envArgs string, extraConfigPaths ...string) (*
 
+  // Set the final value of old range paramaters too for consistency
+
+	if n.IPAM.Range != "" {
+    idx := len(n.IPAM.IPRanges)-1
+    n.IPAM.Range = n.IPAM.IPRanges[idx].Range
+    n.IPAM.RangeStart = n.IPAM.IPRanges[idx].RangeStart
+    n.IPAM.RangeEnd = n.IPAM.IPRanges[idx].RangeEnd
+	}
```
This code is not indented correctly. 

Please run a formatter in your code.

Ayush21298 replied:
Thanks. Done :+1:

maiqueb reviewed (changes requested):
My biggest issue with this PR is how the commits are structured - some of them should be squashed together, since they tell a comprehensive story, making it simpler for the reviewer to understand the full-scope of the change.

First of, I think you should just re-write the entire story of this PR, having rebased from latest master first. 

Then, once the "conflict resolution" commits are gone - and you're forced to not add code to places that no longer exist - we can properly review this.

maiqueb commented:
> @maiqueb Thank you so much for your review :)
> 
> I intended to rewrite the history properly once I got review for the overall structure of the code, since the changes were made iteratively due to some other PRs which had major changes in the code-base.
> 
> Thank you so much for putting an effort to review the PR commit-wise irrespective of the presence of hindrances like conflict resolution commits etc.
> 
> I'll update the code-base to address your review comments and then rewrite the history so that the updates can be included/excluded in the proper commits where they should be placed :)

Perfect! Just let me know when this becomes ready to be looked at.

Ayush21298 commented:
> Perfect! Just let me know when this becomes ready to be looked at.

Sure. Thanks you so much again for your review :)

Ayush21298 commented on e2e/e2e_test.go:
```diff
@@ -106,9 +112,84 @@ var _ = Describe("Whereabouts functionality", func() {
+			})
+
+			It("allocates a single pod within the correct IP ranges", func() {
+				By("checking pod IP is within whereabouts IPAM ranges")
```
Same as above.
Should I remove it or let it be as it is o keep the coding style consistent with other tests ?

Ayush21298 commented on e2e/e2e_test.go:
```diff
@@ -475,12 +556,14 @@ func macvlanNetworkWithWhereaboutsIPAMNetwork(networkName string, namespaceName
         ]
-    }`, ipRange)
+    }`, ipRange, createIPRanges(ipRanges))
+	fmt.Printf("Ayush test macvlanConfig: %s", macvlanConfig)
```
Done :+1:

Ayush21298 commented on pkg/storage/etcd.go:
```diff
@@ -163,27 +163,28 @@ func (p *ETCDIPPool) Update(ctx context.Context, reservations []types.IPReservat
 
 // IPManagement manages ip allocation and deallocation from a storage perspective
-func IPManagementEtcd(ctx context.Context, mode int, ipamConf types.IPAMConfig, containerID string, podRef string) (net.IPNet, error) {
+func IPManagementEtcd(ctx context.Context, mode int, ipamConf types.IPAMConfig, containerID string, podRef string) ([]net.IPNet, error) {
```
It was automatically fixed while squashing the commit and rewriting the history.

Ayush21298 commented on pkg/reconciler/controlloop/pod.go:
```diff
@@ -45,7 +45,7 @@ const (
 )
 
-type garbageCollector func(ctx context.Context, mode int, ipamConf types.IPAMConfig, containerID string, podRef string) (net.IPNet, error)
+type garbageCollector func(ctx context.Context, mode int, ipamConf types.IPAMConfig, containerID string, podRef string) ([]net.IPNet, error)
```
It was automatically fixed while squashing the commit and rewriting the history.

Ayush21298 commented on pkg/types/types.go:
```diff
@@ -36,6 +36,14 @@ type NetConfList struct {
 	Plugins      []*Net `json:"plugins,omitempty"`
 }
 
+type RangeConfiguration struct {
```
Done :+1:

Ayush21298 commented on cmd/whereabouts.go:
```diff
@@ -48,34 +48,36 @@ func cmdAdd(args *skel.CmdArgs) error {
 	switch ipamConf.Datastore {
 	case types.DatastoreETCD:
-		newip, err = storage.IPManagementEtcd(ctx, types.Allocate, *ipamConf, args.ContainerID, getPodRef(args.Args))
+		newips, err = storage.IPManagementEtcd(ctx, types.Allocate, *ipamConf, args.ContainerID, getPodRef(args.Args))
```
It was automatically fixed while squashing the commit and rewriting the history.

Ayush21298 commented on .github/ISSUE_TEMPLATE/bug_report.md:
```diff
@@ -0,0 +1,32 @@
+---
+name: Bug report
```
Done :+1:

Ayush21298 commented on pkg/types/types.go:
```diff
@@ -48,24 +48,24 @@ type RangeConfiguration struct {
-	Type                     string            `json:"type"`
-	Routes                   []*cnitypes.Route `json:"routes"`
-	Addresses                []Address         `json:"addresses,omitempty"`
+	Type                     string               `json:"type"`
```
Done :+1:

Ayush21298 commented on e2e/e2e_test.go:
```diff
@@ -43,19 +43,25 @@ func TestWhereaboutsE2E(t *testing.T) {
+			testNamespace            = "default"
+			ipv4TestRange            = "10.10.0.0/16"
+			testNetworkName          = "wa-nad"
+			testDualStackNetworkName = "wa-dualstack-nad"
```
Done :+1:

Ayush21298 commented on e2e/e2e_test.go:
```diff
@@ -106,9 +112,84 @@ var _ = Describe("Whereabouts functionality", func() {
+			})
+		})
+
+		Context("DualStack test (IPRanges without old range)", func() {
```
Done :+1:

Ayush21298 commented on e2e/retrievers/pod.go:
```diff
@@ -19,28 +19,28 @@ func filterNetworkStatus(
 	podNetStatus, found := pod.Annotations[nettypes.NetworkStatusAnnot]
 	if !found {
-		return "", fmt.Errorf("the pod must feature the `networks-status` annotation")
+		return []string{}, fmt.Errorf("the pod must feature the `networks-status` annotation")
```
Thanks for the suggestion :)
Updated the code to return `nil` instead on empty string :+1:

Ayush21298 commented on pkg/config/config.go:
```diff
@@ -76,20 +76,20 @@ func LoadIPAMConfig(bytes []byte, envArgs string, extraConfigPaths ...string) (*
-    n.IPAM.IPRanges = append(n.IPAM.IPRanges, oldRange)
-  }
+		n.IPAM.IPRanges = append(n.IPAM.IPRanges, oldRange)
+	}
```
Done :+1:

Ayush21298 commented:
Oh I can't request review from both @dougbtv and @maiqueb at the same time :thinking:

Ayush21298 commented:
@maiqueb I have squashed the code, re-based and rewritten the commit history :)

Most of your comments were addressed by this and for rest of them I did the appropriate changes in the code.

Please review it when you are available :)

Secondly, if you want, we can squash the commits more, but that will make the commit a bit complex for reviewing. Let me know if you want me to do that. Thanks :)

maiqueb commented on e2e/e2e_test.go:
```diff
@@ -106,9 +106,97 @@ var _ = Describe("Whereabouts functionality", func() {
+					By("checking pod IP is within whereabouts IPAM ranges")
+					secondaryIfaceIPs, err := retrievers.SecondaryIfaceIPValue(pod)
+					Expect(err).NotTo(HaveOccurred())
+					Expect(inRange(dualStackIPv4Range, secondaryIfaceIPs[0])).To(Succeed())
+					Expect(inRange(dualStackIPv6Range, secondaryIfaceIPs[1])).To(Succeed())
```
Please assert the length of the slice before assuming the indexes: `Expect(secondaryIfaceIPs).To(HaveLen(2))`

maiqueb reviewed (changes requested):
This PR is amazingly well written. Reviewing it a pleasure.

Most of what I have are simple nits intending to improve the code quality / readability further.

Thanks again for this. 

IMO, this is **very** close to being merged.

maiqueb commented on pkg/config/config.go:
```diff
@@ -76,33 +76,57 @@ func LoadIPAMConfig(bytes []byte, envArgs string, extraConfigPaths ...string) (*
-			firstip = netutils.ParseIPSloppy(firstip.Mask(ipNet.Mask).String()) // if range_start is not net then pick the first network address
-			n.IPAM.RangeStart = firstip
+
+		n.IPAM.IPRanges = append(n.IPAM.IPRanges, oldRange)
```
(this is an honest question... I am unsure if this would be a bad idea or not).

Ayush21298 commented on cmd/whereabouts_test.go:
```diff
@@ -531,6 +531,176 @@ var _ = Describe("Whereabouts operations", func() {
+
+		ipamConf, cniVersion, err := config.LoadIPAMConfig([]byte(conf), cniArgs(podNamespace, podName), "")
+		Expect(err).NotTo(HaveOccurred())
+		k8sClient = newK8sIPAM(args.ContainerID, ipamConf, fakek8sclient.NewSimpleClientset(), fake.NewSimpleClientset(ipPool(ipamConf.Range, podNamespace)))
```
Updated :)

Ayush21298 commented on cmd/whereabouts_test.go:
```diff
@@ -531,6 +531,176 @@ var _ = Describe("Whereabouts operations", func() {
+        "ipRanges": [{
+			    "range": "192.168.10.1/24"
+        }]
+			}
```
Oh I didn't realize that `go fmt` did not format `Sprintf`.
Updated :)

Ayush21298 commented on cmd/whereabouts_test.go:
```diff
@@ -531,6 +531,176 @@ var _ = Describe("Whereabouts operations", func() {
+
+		ipamConf, cniVersion, err := config.LoadIPAMConfig([]byte(conf), cniArgs(podNamespace, podName), "")
+		Expect(err).NotTo(HaveOccurred())
+		k8sClient = newK8sIPAM(args.ContainerID, ipamConf, fakek8sclient.NewSimpleClientset(), fake.NewSimpleClientset(ipPool(ipamConf.Range, podNamespace)))
```
Same as above :)

Ayush21298 commented on cmd/whereabouts_test.go:
```diff
@@ -531,6 +531,176 @@ var _ = Describe("Whereabouts operations", func() {
+
+		// Gomega is cranky about slices with different caps
+
+		Expect(result.IPs[0].Address).To(Equal(mustCIDR("192.168.10.1/24")))
+		Expect(result.IPs[1].Address).To(Equal(mustCIDR("abcd::1/64")))
```
> Should we assert the length of `result.IPs` before assuming the indexes ?

Done :+1: 



> Or better yet:
> 
> ```go
> Expect(result.IPs).To(
>     ConsistOf(
>         mustCIDR("192.168.10.1/24"), 
>         mustCIDR("abcd::1/64")))
> ```

We can't use this because `results.IPs` is of type `[]*current.IPConfig` whereas `mustCIDR` if of type `net.IPNet`.
Additionally, I made a new function `ipConfigCIDR` which is of type `current.IPConfig`.
But this can't be used with `results.IPs` without creating a separate matcher. So I just asserted the lendth

maiqueb replied:
Fair enough. I'll sort that out once I can spare the time. 

Thanks.

Ayush21298 commented on cmd/whereabouts_test.go:
```diff
@@ -531,6 +531,176 @@ var _ = Describe("Whereabouts operations", func() {
+
+		// Gomega is cranky about slices with different caps
+
+		Expect(result.IPs[0].Address).To(Equal(mustCIDR("192.168.10.1/24")))
```
Same as above :)

Ayush21298 commented on cmd/whereabouts_test.go:
```diff
@@ -531,6 +531,176 @@ var _ = Describe("Whereabouts operations", func() {
+			  %s,
+        "ipRanges": [{
+			    "range": "192.168.10.1/24"
+        }],
```
Same as above :)

Ayush21298 commented on cmd/whereabouts_test.go:
```diff
@@ -531,6 +531,176 @@ var _ = Describe("Whereabouts operations", func() {
+		// Gomega is cranky about slices with different caps
+
+		Expect(result.IPs[0].Address).To(Equal(mustCIDR("192.168.10.1/24")))
+		Expect(result.IPs[1].Address).To(Equal(mustCIDR("abcd::1/64")))
```
Same as above :)

Ayush21298 commented on e2e/e2e_test.go:
```diff
@@ -106,9 +106,97 @@ var _ = Describe("Whereabouts functionality", func() {
+					const dualstackPodName = "whereabouts-dualstack-test"
+					var err error
+
+					netAttachDefDualStack = macvlanNetworkWithWhereaboutsIPAMNetwork(testDualStackNetworkName, testNamespace, "", testIPRangesDualStack)
```
Done :+1:

Ayush21298 commented:
@maiqueb I have added a separate new commit with the changes for addressing the review comment for the ease in review. Once I get a green light from you, I'll add the respective sections in appropriate commits.
Thanks :)

maiqueb commented on cmd/whereabouts_test.go:
```diff
@@ -1341,15 +1359,15 @@ users:
 `)
 }
 
-func ipPool(ipRange string, namespace string, podReferences ...string) *v1alpha1.IPPool {
```
I still think it is the responsibility of the caller to pass the `ipRange`. It is a code smell that we pass it the whole configuration when this just uses the range of the configuration. Especially one that assumes a particular index.

I would leave the function signature as it was, and pass it from the the test the range from which you want to build an `IPAMConfig` from.

Ayush21298 replied:
Thanks :)
I rolled back my changes in `ipPool` function and instead updated the test cases to pass the the range from the 1st element of `IPRanges` instead to "oldRange" (since it was removed).

maiqueb commented on pkg/config/config.go:
```diff
@@ -86,7 +86,7 @@ func LoadIPAMConfig(bytes []byte, envArgs string, extraConfigPaths ...string) (*
 			RangeEnd:   n.IPAM.RangeEnd,
 		}
 
-		n.IPAM.IPRanges = append(n.IPAM.IPRanges, oldRange)
```
I don't think you actually deleted the "oldRange" from the configuration (below, lines 121 to 129). 

Honestly I'm unsure if we want to do that. Let's ask @dougbtv 's opinion.

Ayush21298 replied:
I deleted "oldRange" completely right after copying it as the first element of `IPRanges` and updated the corresponding tests too :)

maiqueb reviewed (changes requested):
IMO, you should just leave the `ipPool` function signature as it was before. 

Then you can squash this last commit, and I would be willing to proceed with a merge.

Let's also wait for @dougbtv 's feedback / opinion on how he things we should enforce consistency between the "old" single range and the array (by deleting from the old, or by copying from the array to the old attribute).

Ayush21298 commented:
@maiqueb Thanks for your opinions :)
I have removed the "old range" completely as we discussed in the meeting, addressed all your comments and added the related fixes in their respective commits. Please review it when you are available :)

maiqueb commented on e2e/e2e_test.go:
```diff
@@ -106,9 +106,107 @@ var _ = Describe("Whereabouts functionality", func() {
 			It("allocates a single pod within the correct IP range", func() {
 				By("checking pod IP is within whereabouts IPAM range")
-				secondaryIfaceIP, err := retrievers.SecondaryIfaceIPValue(pod)
+				secondaryIfaceIPs, err := retrievers.SecondaryIfaceIPValue(pod)
```
please assert these are not empty.

Ayush21298 replied:
Thanks. Updated it :)

maiqueb reviewed (changes requested):
I just have minor comments requesting for a further assertion in the tests, which will hopefully pinpoint the test failure should it happen.

I think @dougbtv can start to have a look into this never-the-less. I'm OK with the change.

Ayush21298 commented on cmd/whereabouts_test.go:
```diff
@@ -531,6 +531,194 @@ var _ = Describe("Whereabouts operations", func() {
+		}
+
+		ipamConf, cniVersion, err := config.LoadIPAMConfig([]byte(conf), cniArgs(podNamespace, podName), "")
+		Expect(err).NotTo(HaveOccurred())
```
Thanks. Updated it in all the places :)

Ayush21298 commented on cmd/whereabouts_test.go:
```diff
@@ -531,6 +531,194 @@ var _ = Describe("Whereabouts operations", func() {
+			ipamConf,
+			fakek8sclient.NewSimpleClientset(),
+			fake.NewSimpleClientset(
+				ipPool(ipamConf.IPRanges[0].Range, podNamespace)))
```
Thanks. Updated it in all the places :)

Ayush21298 commented on cmd/whereabouts_test.go:
```diff
@@ -531,6 +531,194 @@ var _ = Describe("Whereabouts operations", func() {
+			ipamConf,
+			fakek8sclient.NewSimpleClientset(),
+			fake.NewSimpleClientset(
+				ipPool(ipamConf.IPRanges[0].Range, podNamespace)))
```
Thanks. Updated it in all the places :)

Ayush21298 commented on cmd/whereabouts_test.go:
```diff
@@ -531,6 +531,194 @@ var _ = Describe("Whereabouts operations", func() {
+			ipamConf,
+			fakek8sclient.NewSimpleClientset(),
+			fake.NewSimpleClientset(
+				ipPool(ipamConf.IPRanges[0].Range, podNamespace)))
```
Yes, you're absolutely correct. The other one is lazily created :)

maiqueb replied:
I think we should be consistent here ... either always lazy, or always create the pools.

Am I missing something ?

Ayush21298 replied:
As far as I remember, I didn't find any tests for lazy IP Pool creation, so though of initializing one lazily to test that.
But nevertheless, since it is out of scope of this PR (and also makes the test case confusing), I updated the code to create both the IP Pools :)
Thanks for pointing it out ^^

Ayush21298 commented:
@maiqueb Thanks for your review again :)
Updated the code in respective commits to address the comments ^^

maiqueb reviewed (approved):
Thank you for the feature, and for your patience.

devenderkamble2101 commented:
which release dual stack feature is available.

Ayush21298 commented:
> which release dual stack feature is available.

Hopefully from the next release :)

@dougbtv and @maiqueb can tell better ^^

dougbtv reviewed (approved):
LGTM! Incredible work. Thank you so much Ayush. I ran through a few last minute QE manual tests in my lab to check vs. previous version and backwards compatibility and it all looks great. Excellent!

namtab00 commented:
does this allow, in any way, assigning a "floating/virtual" IP to multiple pods?

I'm trying to listen to a single external UDP stream from multiple pods that have a multus mapped secondary NIC using a single master host NIC.
AFAIK having a single IP assigned to multiple pods is impossible, but I'd gladly stand corrected...
